<center>
<img src="https://habrastorage.org/files/fd4/502/43d/fd450243dd604b81b9713213a247aa20.jpg">
    
## [mlcourse.ai](mlcourse.ai) – Open Machine Learning Course 

<center>Author: [Yury Kashnitskiy](http://yorko.github.io) <br>
Translated and edited by [Sergey Isaev](https://www.linkedin.com/in/isvforall/), [Artem Trunov](https://www.linkedin.com/in/datamove/), [Anastasia Manokhina](https://www.linkedin.com/in/anastasiamanokhina/), and [Yuanyuan Pao](https://www.linkedin.com/in/yuanyuanpao/) <br>All content is distributed under the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license.

# <center> Assignment #1 (demo)
## <center>  Exploratory data analysis with Pandas


**In this task you should use Pandas to answer a few questions about the [Adult](https://archive.ics.uci.edu/ml/datasets/Adult) dataset. (You don't have to download the data – it's already here). Choose the answers in the [web-form](https://docs.google.com/forms/d/1uY7MpI2trKx6FLWZte0uVh3ULV4Cm_tDud0VDFGCOKg). This is a demo version of an assignment, so by submitting the form, you'll see a link to the solution .ipynb file.**

Unique values of all features (for more information, please see the links above):
- `age`: continuous.
- `workclass`: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked.
- `fnlwgt`: continuous.
- `education`: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool.
- `education-num`: continuous.
- `marital-status`: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse.
- `occupation`: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces.
- `relationship`: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried.
- `race`: White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other, Black.
- `sex`: Female, Male.
- `capital-gain`: continuous.
- `capital-loss`: continuous.
- `hours-per-week`: continuous.
- `native-country`: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.   
- `salary`: >50K,<=50K

In [40]:
import pandas as pd
import numpy as np

In [41]:
data = pd.read_csv('../input/adult.data.csv')
data.head()

age         workclass  fnlwgt  ...   hours-per-week  native-country salary
0   39         State-gov   77516  ...               40   United-States  <=50K
1   50  Self-emp-not-inc   83311  ...               13   United-States  <=50K
2   38           Private  215646  ...               40   United-States  <=50K
3   53           Private  234721  ...               40   United-States  <=50K
4   28           Private  338409  ...               40            Cuba  <=50K

[5 rows x 15 columns]

**1. How many men and women (*sex* feature) are represented in this dataset?** 

In [42]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
age               32561 non-null int64
workclass         32561 non-null object
fnlwgt            32561 non-null int64
education         32561 non-null object
education-num     32561 non-null int64
marital-status    32561 non-null object
occupation        32561 non-null object
relationship      32561 non-null object
race              32561 non-null object
sex               32561 non-null object
capital-gain      32561 non-null int64
capital-loss      32561 non-null int64
hours-per-week    32561 non-null int64
native-country    32561 non-null object
salary            32561 non-null object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [43]:
# You code here
print('Number of men: ',data['sex'].value_counts()[0])
print('Number of women: ',data['sex'].value_counts()[1])

Number of men:  21790
Number of women:  10771


**2. What is the average age (*age* feature) of women?**

In [44]:
# You code here
x = data.groupby('sex').mean()['age'][0]
print('Average age of women : ',x)

Average age of women :  36.85823043357163


**3. What is the percentage of German citizens (*native-country* feature)?**

In [45]:
# You code here
german_cit = data['native-country'].value_counts()[4]
print(' percentage of German citizens: {:.3f} '.format(german_cit/len(data)))

 percentage of German citizens: 0.004 


**4-5. What are the mean and standard deviation of age for those who earn more than 50K per year (*salary* feature) and those who earn less than 50K per year? **

In [46]:
# You code here
data['salary'].unique()

array(['<=50K', '>50K'], dtype=object)

In [47]:
data.groupby('salary').agg(['mean','std'])['age']

mean        std
salary                      
<=50K   36.783738  14.020088
>50K    44.249841  10.519028

**6. Is it true that people who earn more than 50K have at least high school education? (*education – Bachelors, Prof-school, Assoc-acdm, Assoc-voc, Masters* or *Doctorate* feature)**

In [48]:
data['education'].unique()

array(['Bachelors', 'HS-grad', '11th', 'Masters', '9th', 'Some-college',
       'Assoc-acdm', 'Assoc-voc', '7th-8th', 'Doctorate', 'Prof-school',
       '5th-6th', '10th', '1st-4th', 'Preschool', '12th'], dtype=object)

In [49]:
# You code here
data.groupby(['salary','education']).count()['education-num']

salary  education   
<=50K   10th             871
        11th            1115
        12th             400
        1st-4th          162
        5th-6th          317
        7th-8th          606
        9th              487
        Assoc-acdm       802
        Assoc-voc       1021
        Bachelors       3134
        Doctorate        107
        HS-grad         8826
        Masters          764
        Preschool         51
        Prof-school      153
        Some-college    5904
>50K    10th              62
        11th              60
        12th              33
        1st-4th            6
        5th-6th           16
        7th-8th           40
        9th               27
        Assoc-acdm       265
        Assoc-voc        361
        Bachelors       2221
        Doctorate        306
        HS-grad         1675
        Masters          959
        Prof-school      423
        Some-college    1387
Name: education-num, dtype: int64

**No it is not true that people who earn more than 50K have at least high school education **

**7. Display age statistics for each race (*race* feature) and each gender (*sex* feature). Use *groupby()* and *describe()*. Find the maximum age of men of *Amer-Indian-Eskimo* race.**

In [50]:
# You code here
x = data.groupby(['race','sex']).describe()['age']
x

count       mean  ...     75%   max
race               sex                         ...              
Amer-Indian-Eskimo Female    119.0  37.117647  ...   46.00  80.0
                   Male      192.0  37.208333  ...   45.00  82.0
Asian-Pac-Islander Female    346.0  35.089595  ...   43.75  75.0
                   Male      693.0  39.073593  ...   46.00  90.0
Black              Female   1555.0  37.854019  ...   46.00  90.0
                   Male     1569.0  37.682600  ...   46.00  90.0
Other              Female    109.0  31.678899  ...   39.00  74.0
                   Male      162.0  34.654321  ...   42.00  77.0
White              Female   8642.0  36.811618  ...   46.00  90.0
                   Male    19174.0  39.652498  ...   49.00  90.0

[10 rows x 8 columns]

In [51]:
print('Maximum age of men of Amer-Indian-Eskimo:  ',x.loc['Amer-Indian-Eskimo'].loc['Male','max'])

Maximum age of men of Amer-Indian-Eskimo:   82.0


**8. Among whom is the proportion of those who earn a lot (>50K) greater: married or single men (*marital-status* feature)? Consider as married those who have a *marital-status* starting with *Married* (Married-civ-spouse, Married-spouse-absent or Married-AF-spouse), the rest are considered bachelors.**

In [52]:
# You code here
data['married'] = data['marital-status'].apply(lambda x:1 if x.split('-')[0]=='Married' else 0)
data.head()

age         workclass  fnlwgt   ...   native-country  salary married
0   39         State-gov   77516   ...    United-States   <=50K       0
1   50  Self-emp-not-inc   83311   ...    United-States   <=50K       1
2   38           Private  215646   ...    United-States   <=50K       0
3   53           Private  234721   ...    United-States   <=50K       1
4   28           Private  338409   ...             Cuba   <=50K       1

[5 rows x 16 columns]

In [53]:
count = data.groupby(['salary','sex'])['married'].value_counts()
count

salary  sex     married
<=50K   Female  0          8487
                1          1105
        Male    1          7576
                0          7552
>50K    Female  1           771
                0           408
        Male    1          5965
                0           697
Name: married, dtype: int64

In [54]:
percent = count.groupby(level=[0,1]).apply(lambda x: x/(x.sum()))
percent

salary  sex     married
<=50K   Female  0          0.884800
                1          0.115200
        Male    1          0.500793
                0          0.499207
>50K    Female  1          0.653944
                0          0.346056
        Male    1          0.895377
                0          0.104623
Name: married, dtype: float64

**So the answer to the above question is more married men(89.5%) are earning >50k compared to single men(10.4%)**

**9. What is the maximum number of hours a person works per week (*hours-per-week* feature)? How many people work such a number of hours, and what is the percentage of those who earn a lot (>50K) among them?**

In [55]:
# You code here
print('Maximum no. of hours a person works per week: ',data['hours-per-week'].max())

Maximum no. of hours a person works per week:  99


In [56]:
print('Number of people who work for 99 hours per weeek: ',len(data[data['hours-per-week']==99]))

Number of people who work for 99 hours per weeek:  85


In [57]:
count_sal = data[data['hours-per-week']==99].groupby('salary').count()['hours-per-week']
count_sal

salary
<=50K    60
>50K     25
Name: hours-per-week, dtype: int64

In [58]:
count_sal/count_sal.sum()

salary
<=50K    0.705882
>50K     0.294118
Name: hours-per-week, dtype: float64

**From the above solution we can infer that 29.4% of people who works for 99 hours per week get >50k salary **

**10. Count the average time of work (*hours-per-week*) for those who earn a little and a lot (*salary*) for each country (*native-country*). What will these be for Japan?**

In [59]:
# You code here
data.groupby(['salary','native-country']).mean()['hours-per-week']

salary  native-country            
<=50K   ?                             40.164760
        Cambodia                      41.416667
        Canada                        37.914634
        China                         37.381818
        Columbia                      38.684211
        Cuba                          37.985714
        Dominican-Republic            42.338235
        Ecuador                       38.041667
        El-Salvador                   36.030928
        England                       40.483333
        France                        41.058824
        Germany                       39.139785
        Greece                        41.809524
        Guatemala                     39.360656
        Haiti                         36.325000
        Holand-Netherlands            40.000000
        Honduras                      34.333333
        Hong                          39.142857
        Hungary                       31.300000
        India                         38.233333
     

Persons who earn <=50k in Japan work for **41** hours on an average and the people who earn >50k in Japan work for **47.9** hours on an average